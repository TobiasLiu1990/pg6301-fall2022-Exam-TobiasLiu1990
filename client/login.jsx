import * as React from "react";
import { useState } from "react";
import { useNavigate } from "react-router-dom";
import { FormInput } from "./lib/formInput";
import {fetchJSON} from "./lib/fetchJson";

export function Login() {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const navigate = useNavigate();

  async function handleSubmit(e) {
    e.preventDefault();
    //POST Login
    const res = await fetchJSON("/api/login", {
      method: "post",
      json: ({ username, password }),
    });

    if (!res.ok) {
      alert("Unauthorized");
    }
    if (res.ok) {
      navigate("/");
    }
  }

  function handleSubmitBack(e) {
    e.preventDefault();
    navigate("/");
  }

  return (
    <div>
      <form onSubmit={handleSubmitBack}>
        <button>Go back to start page</button>
      </form>

      <hr></hr>

      <h1>Please login!</h1>
      <form onSubmit={handleSubmit}>
        <FormInput
          label={"Username: "}
          value={username}
          onChangeValue={setUsername}
        />
        <FormInput
          label={"Password: "}
          value={password}
          onChangeValue={setPassword}
        />
        <button>Log in</button>
      </form>
    </div>
  );
}

export function LoginForTest({ loginApi }) {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const navigate = useNavigate();

  loginApi.fakeLogin({ username, password }); //For test

  function handleSubmitBack(e) {
    e.preventDefault();
    navigate("/");
  }

  function handleSubmit(e) {
    e.preventDefault();
  }

  return (
    <div>
      <form onSubmit={handleSubmitBack}>
        <button>Go back to start page</button>
      </form>

      <hr></hr>

      <h1>Please login!</h1>

      <form onSubmit={handleSubmit}>
        <FormInput
          label={"Username: "}
          value={username}
          onChangeValue={setUsername}
        />
        <FormInput
          label={"Password: "}
          value={password}
          onChangeValue={setPassword}
        />
        <button>Log in</button>
      </form>
    </div>
  );
}