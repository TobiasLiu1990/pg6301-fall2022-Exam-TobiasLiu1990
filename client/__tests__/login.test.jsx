import ReactDOM from "react-dom";
import React from "react";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";
import { Login, LoginForTest } from "../login";
import { queryAllByAltText } from "@testing-library/react";

const user = [
  {
    role: "admin",
    username: "administrator",
    fullName: "mr. admin",
    password: "secret",
  },
];

describe("login component", () => {
  it("should show login form", async () => {
    const element = document.createElement("div");
    await act(async () =>
      ReactDOM.render(
        <MemoryRouter>
          <Login />
        </MemoryRouter>,
        element
      )
    );
    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should simulate login form", async () => {
    const fakeLogin = jest.fn();
    const element = document.createElement("div");

    await act(async () =>
      ReactDOM.render(
        <MemoryRouter>
          <LoginForTest loginApi={{ fakeLogin }} />
        </MemoryRouter>,
        element
      )
    );

    Simulate.change(
      element.querySelector("form:nth-of-type(2) div:nth-of-type(1) input"),
      {
        target: { value: "admin" },
      }
    );

    Simulate.change(
      element.querySelector("form:nth-of-type(2) div:nth-of-type(2) input"),
      {
        target: { value: "password" },
      }
    );

    Simulate.submit(element.querySelector("form"));

    expect(fakeLogin).toBeCalledWith({
      username: "admin",
      password: "password",
    });
  });
});
