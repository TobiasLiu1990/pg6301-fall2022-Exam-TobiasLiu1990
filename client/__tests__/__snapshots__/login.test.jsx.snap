// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login component should show login form 1`] = `"<div><form><button>Go back to start page</button></form><hr><h1>Please login!</h1><form><div><label><strong>Username: </strong> <input value=""></label><br></div><label><strong>Password: </strong><input type="password" value=""></label><button>Log in</button></form></div>"`;
